Include some LLM response metadata in `LLMMessage`

AdeelH opened the pull request:
Currently includes token usage and stop reason.

AdeelH committed:
include some LLM response metadata in LLMMessage

AdeelH committed:
update tests

element84-infra commented:
### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/geoai/pr-checks/134ce38f-842b-4f49-940c-ae4ba1c554bf?refs=252f245a5f79bd0703bebffb00288efa452a1794&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/geoai/pr-checks/134ce38f-842b-4f49-940c-ae4ba1c554bf/license?refs=252f245a5f79bd0703bebffb00288efa452a1794&source=prChecksComment)

AdeelH commented on src/e84_geoai_common/llm/core/llm.py:
```diff
@@ -79,6 +89,7 @@ class LLMMessage(BaseModel):
 
     role: Literal["assistant", "user"] = "user"
     content: str | Sequence[LLMMessageContentType]
+    metadata: LLMMessageMetadata | None = Field(default=None, exclude=True)
```
`exclude=True` ensures that this field is not included when this message is serialized in subsequent calls to the LLM.

AdeelH replied:
Update: `exclude=True` was unnecessary and has now been removed.

AdeelH committed:
remove exclude=True

We are not directly passing LLMMessage to the API, so this is not needed.

jasongilman commented on src/e84_geoai_common/llm/core/llm.py:
```diff
@@ -68,6 +68,16 @@ class CachePointContent(BaseModel):
 )
 
 
+class LLMMessageMetadata(BaseModel):
```
I'm imagining an LLM Message with metadata that comes from the user. Would that ever be the case? It seems like this metadata would only be present if the role is "assistant". I think that's probably ok though.

jasongilman commented on src/e84_geoai_common/llm/core/llm.py:
```diff
@@ -68,6 +68,16 @@ class CachePointContent(BaseModel):
+
+    input_tokens: int
+    output_tokens: int
+    stop_reason: Literal["end_turn", "max_tokens", "stop_sequence", "tool_use"]
```
What does `end_turn` mean? Is that different from `tool_use`?

AdeelH replied:
Yes. `end_turn` is what most LLM APIs normally return to indicate that the LLM is done with its message. `tool_use` means it stopped for a tool call and is expecting results.
https://docs.anthropic.com/en/api/handling-stop-reasons

AdeelH committed:
don't suppress reportIncompatibleVariableOverride

AdeelH committed:
refactor LLMMessage

- Define LLMAssistantMessage and LLMUserMessage and move metadata to LLMAssistantMessage.
- Rename the metadata class to LLMResponseMetadata

AdeelH committed:
unbreak tests with live LLM

- The dummy input/output token counts were not compatible with a real LLM client. We now exclude token counts from the comparison.
- Nova's response parsing has been made more consistent with the other models'.
